### PR TITLE
Add `inference_mode` flag to Trainer

### LINKED ISSUE #14497: Need an option for evaluation not in the inference_mode.
## 🚀 Feature

An option for the trainer run evaluation not in the inference_mode.

### Motivation

In some scenerios, gradients are used even when evaluating the model. The recent PR (https://github.com/Lightning-AI/lightning/pull/12715) adds supports for the new PyTorch API `inference_mode` and enables it by default. However, unlike `torch.no_grad` that one can simply set `torch.enable_grad` to overwrite it locally, `inference_mode` do not provide a convenient way to quit it locally. This is because all tensors are marked and forbidden for furthur gradient-related usage. People have to clone tensors to make code working. Therefore, I suggest that pytorch-lightning provide a way to let users to choose whether to use the inference mode.

### Pitch

Add an option to the Trainer class that allows users to decide whether to use the inference mode or not.

### Additional context

From the pytorch's document https://pytorch.org/cppdocs/notes/inference_mode.html:
> Inference tensors:
> ...
> are immutable outside InferenceMode. So an error will be raised if you try to: - mutate their data outside InferenceMode. - mutate them into requires_grad=True outside InferenceMode. To work around you can make a clone outside InferenceMode to get a normal tensor before mutating.

cc @justusschock @kaushikb11 @awaelchli @borda @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by rschireman**
## What does this PR do?

Fixes #14497 by adding a flag to the trainer to control the behavior of autograd in inference mode, such as during model testing. 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -388,6 +389,9 @@ def __init__(
                 and smaller datasets reload when running out of their data. In 'min_size' mode, all the datasets
                 reload when reaching the minimum length of datasets.
                 Default: ``"max_size_cycle"``.
+
+            inference_grad_mode: Controls if autograd is used in the inference phase, such as during model testing.
+                Default: ``False``.
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:394]:
```suggestion
```
not required.

File: src/pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -162,6 +162,7 @@ def __init__(
         amp_level: Optional[str] = None,
         move_metrics_to_cpu: bool = False,
         multiple_trainloader_mode: str = "max_size_cycle",
+        inference_grad_mode: Optional[bool] = False,
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:165]:
```suggestion
        inference_grad_mode: bool = False,
```
can be just bool

File: src/pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2228,15 +2236,20 @@ def configure_optimizers(self):
 
 
 @contextmanager
-def _evaluation_context(accelerator: Accelerator) -> Generator:
+def _evaluation_context(accelerator: Accelerator, grad_mode: Optional[bool] = False) -> Generator:
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:2239]:
```suggestion
def _evaluation_context(accelerator: Accelerator, inference_grad_mode: bool = False) -> Generator:
```

File: src/pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -2228,15 +2236,20 @@ def configure_optimizers(self):
...
     context_manager_class = (
         torch.inference_mode
         if not (dist.is_available() and dist.is_initialized() and dist.get_backend() == "gloo")
         and not isinstance(accelerator, HPUAccelerator)
         and not isinstance(accelerator, TPUAccelerator)
         else torch.no_grad
     )
+
+    if grad_mode:
+        context_manager_class = torch.enable_grad
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:2252]:
```suggestion
    if not inference_grad_mode:
        context_manager_class = torch.no_grad
```

File: src/pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -388,6 +389,9 @@ def __init__(
                 and smaller datasets reload when running out of their data. In 'min_size' mode, all the datasets
                 reload when reaching the minimum length of datasets.
                 Default: ``"max_size_cycle"``.
+
+            inference_grad_mode: Controls if autograd is used in the inference phase, such as during model testing.
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/trainer.py:393]: default should be True

File: tests/tests_pytorch/trainer/flags/test_inference_grad_mode.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        assert torch.is_inference_mode_enabled
+        return super().on_test_epoch_start()
+
+
+def test_inference_grad_mode():
+    """Testing overwriting trainer arguments."""
+    trainer = Trainer(logger=False, inference_grad_mode=True)
+    trainer.test(_BoringModelForEnableGrad())
+    trainer = Trainer(logger=False, inference_grad_mode=False)
+    trainer.test(_BoringModelForNoGrad())
```
Comment (rohitgr7) [inline tests/tests_pytorch/trainer/flags/test_inference_grad_mode.py:40]: use `fast_dev_run`?

File: src/pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (rschireman) [reply src/pytorch_lightning/trainer/trainer.py:393]: setting `inference_grad_mode=True` as the default causes `def test_evaluation_move_metrics_to_cpu_and_outputs(tmpdir, accelerator):` to fail when combined with:     
```
if not inference_grad_mode:
        context_manager_class = torch.no_grad
```

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/common/trainer.rst [hunk 7]
```diff
@@ -1525,6 +1525,39 @@ Whether to enable or disable the model summarization. Defaults to True.
 
     trainer = Trainer(enable_model_summary=True, callbacks=[ModelSummary(max_depth=-1)])
 
+
+eval_inference_mode
```
Comment (carmocca) [inline docs/source-pytorch/common/trainer.rst:1529]: I would suggest dropping the `eval_` prefix.

Comment (awaelchli) [reply docs/source-pytorch/common/trainer.rst:1529]: Agree, inference describes all three stages in our Trainer: validate, test, predict.

File: docs/source-pytorch/common/trainer.rst [hunk 8]
```diff
@@ -1525,6 +1525,39 @@ Whether to enable or disable the model summarization. Defaults to True.
 
     trainer = Trainer(enable_model_summary=True, callbacks=[ModelSummary(max_depth=-1)])
 
+
+inference_mode
+^^^^^^^^^^^^^^^^^^^
```
Comment (Borda) [inline docs/source-pytorch/common/trainer.rst:1530]:
```suggestion
inference_mode
^^^^^^^^^^^^^^
```

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]
